model/status_diff.go

```diff
@@ -54,45 +55,60 @@ var (
 
 // StatusDiffBuilds takes two builds and returns a diff of their results
 // for easy comparison and analysis.
-func StatusDiffBuilds(original, patch *build.Build) BuildStatusDiff {
+func StatusDiffBuilds(original, patch *build.Build) (BuildStatusDiff, error) {
 	// return an empty diff if one of builds is nonexistant
 	// this is likely to occur after adding a new buildvariant or task
 	if original == nil || patch == nil {
-		return BuildStatusDiff{}
+		return BuildStatusDiff{}, nil
 	}
 
 	diff := BuildStatusDiff{
 		Name: original.DisplayName,
 		Diff: StatusDiff{original.Status, patch.Status},
 	}
 
-	// build maps of task statuses, for matching
-	originalTasks := make(map[string]build.TaskCache)
-	for _, task := range original.Tasks {
-		originalTasks[task.DisplayName] = task
+	tasks, err := task.FindAll(
+		task.ByBuildIds([]string{original.Id, patch.Id}).WithFields(task.BuildIdKey, task.StatusKey, task.DetailsKey, task.DisplayNameKey))
+	if err != nil {
+		return BuildStatusDiff{}, errors.Wrap(err, "can't get tasks")
+	}
+
+	// build maps of tasks, for matching
+	originalTaskMap := make(map[string]task.Task)
+	patchTaskMap := make(map[string]task.Task)
+	for _, t := range tasks {
+		if t.BuildId == patch.Id {
+			patchTaskMap[t.Id] = t
```

Comment:
Why don't we use t.DisplayName to be consistent with the other case?

Reply:
later (on line 90) we pull the task out of the map by task id.

Reply:
bu patchTask has display name, why don't we just get the task using the display name (the way we do with original task map)?

Reply:
my thinking was that we're pulling patch tasks by their id in the task cache, and then comparing the task results with the same task in the base build (which will share a display name).

If we were to iterate over _all the tasks_ in the patch we would also end up with tasks that aren't in the build cache (execution tasks) and they'd be out of order. I'm honestly not sure whether this would break something (I didn't fully investigate).